source/contentprovider.d

```diff
@@ -105,58 +108,63 @@ class ContentProvider
+		enforce(language in langWithPath, "Language hasn't been seen before.");
+		string fileName = buildPath(langWithPath[language], parts[0], parts[1] ~ ".md");
 		import std.stdio;
 		return exists(fileName);
 	}
 
 	this(string contentDirectory)
 	{
 		this.contentDirectory = contentDirectory;
+
 		// parse mustache template file
 		setupMustacheMacros(readText(buildPath(contentDirectory, "template.yml")),
 			mustacheContext_);
-		foreach(string filename; dirEntries(contentDirectory, SpanMode.depth)) {
-			if (isDir(filename))
-				continue;
-			auto parts = filename[contentDirectory.length .. $]
-					.split('/').filter!(x => !x.empty)
-					.array;
-
-			// ignore any top level files
-			if (parts.length == 1) {
-				continue;
-			}
+	}
 
-			// search for language-specific root file
-			auto language = parts[0];
-			if (parts[1] != "index.yml")
-				continue;
-			auto root = Loader(filename).load();
+	public void addLanguages(string directory)
+	{
+		foreach(string filename; dirEntries(directory, SpanMode.shallow))
+			addLanguage(filename);
+	}
 
-			// language meta information
-			enforce("start" in root, "'start' point required in language-specific yaml");
-			enforce(isValidLink(root["start"].as!string, language), "The start page must be formatted as chapter/section");
+	public void addLanguage(string langDirectory)
+	{
+		if (!langDirectory.isDir)
+			return;
 
-			LanguageMeta langMeta;
-			langMeta.start = ChapterAndSection(root["start"].as!string);
+		auto configFile = langDirectory.buildPath("index.yml");
+		if (!configFile.exists)
+			logInfo("Missing index.yml for %s", langDirectory);
 
-			import std.meta : AliasSeq;
-			foreach (attr; AliasSeq!("title", "repo"))
-			{
-				enforce(attr in root, "'" ~ attr ~ "' point required in language-specific yaml");
-				 mixin("langMeta." ~ attr ~ " = root[attr].as!string;");
-			}
-			language_[language] = langMeta;
+		auto language = langDirectory.baseName;
+		auto root = Loader(configFile).load();
 
-		    auto i = 0;
-			foreach (string chapter; root["ordering"])
-			{
-				auto chapterDir = buildPath(filename[0 .. contentDirectory.length],
-					language, chapter);
-				chapter_[language][chapter] = ChapterMeta(i++, "");
-				addChapter(chapter, chapterDir, language);
-			}
+		// store the directory to be able to link to other requests
+		langWithPath[language] = langDirectory;
+
+		// language meta information
+		enforce("start" in root, "'start' point required in language-specific yaml");
+		enforce(isValidLink(root["start"].as!string, language), "The start page must be formatted as chapter/section");
+
+		LanguageMeta langMeta;
+		langMeta.start = ChapterAndSection(root["start"].as!string);
+
+		import std.meta : AliasSeq;
+		foreach (attr; AliasSeq!("title", "repo"))
+		{
+			enforce(attr in root, "'" ~ attr ~ "' point required in language-specific yaml");
+			 mixin("langMeta." ~ attr ~ " = root[attr].as!string;");
+		}
```

Comment:
this might not the best serialization solution, but as long as we don't have many parameters it should work for us.